### PR TITLE
[SILOpt] Handle reborrows of owned phis in CanonicalizeOSSALifetime.

### PR DESCRIPTION
**Pull request by nate-chandler**
When canonicalizing lifetime of an owned phi, update liveness for all reborrows any of whose underlying values are transitively borrows of any of the owned phi's underlying values.

rdar://94346482

### CODE REVIEW

**Comment by nate-chandler** [comment]
@swift-ci please test

**Review by atrick — CHANGES_REQUESTED** [review]
Fantastic code and comments.

My concerns so far (I haven't had time to think of alternatives):

We're repeatedly recomputing phi webs and rediscovering the mapping between all the enclosing scope phis and inner scope phis

The PrunedLiveness abstraction is broken if we allow reborrows. We're either computing SSA liveness or phi-web liveness and that needs to be clear when we handle reborrows. For phi webs, we need to explicitly initialize liveness for all the roots of the phi-web as defs. If we're doing liveness for phi-webs, then we need to make it clear how to use the API correctly.

Along the same lines, we might be visiting uses that aren't dominated by the original def for which we initialized liveness.

We might want a general enclosingScope analysis that maps inner phis to their enclosing scope phis, but I'm not sure we need it yet.

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 1]
```diff
@@ -1586,6 +1586,138 @@ void swift::findTransitiveReborrowBaseValuePairs(
...
+/// this function finds every reborrow
+///     br block(..., %lifetime, ..., %value_orig, ...)
+///                   ^^^^^^^^^
+/// which is a borrow of any of the values %value_orig coming into %value.
+///
+/// Finding these is more complicated than merely looking for guaranteed
+/// operands adjacent to the incoming operands to the owned phi argument and
+/// which are borrows of the value consumed there.  The reason is that they
+/// might not be borrows of that incoming owned value _directly_ but rather
+/// reborrows of some underlying def if the incoming owned value is itself a
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SIL/Utils/OwnershipUtils.cpp]: Reworded in terms of {reaching values of, values reaching} the phi of interest.

File: lib/SIL/Utils/OwnershipUtils.cpp [hunk 2]
```diff
@@ -1586,6 +1586,138 @@ void swift::findTransitiveReborrowBaseValuePairs(
...
+///     br two(%value_1, %lifetime_1)
+///                      ^^^^^^^^^^^
+bool swift::visitReborrowsOfOwnedPhiArgument(
+    SILPhiArgument *arg, function_ref<bool(Operand *)> visitor) {
+  assert(arg->getOwnershipKind() == OwnershipKind::Owned);
+  assert(arg->isPhi());
+
+  // First, collect all the arguments that forward into arg and all the
+  // underlying values which it takes on along different control flow paths.
+  SmallVector<SILPhiArgument *, 4> ownedArguments;
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SIL/Utils/OwnershipUtils.cpp]: Renamed this and other variables, hopefully to be clearer.

File: include/swift/SIL/OwnershipUtils.h [hunk 3]
```diff
@@ -1217,6 +1217,14 @@ void findTransitiveReborrowBaseValuePairs(
     BorrowingOperand initialScopeOperand, SILValue origBaseValue,
     function_ref<void(SILPhiArgument *, SILValue)> visitReborrowBaseValuePair);
 
+/// Visit the reborrows one of whose underlying defs is the same as one of the
+/// underlying defs of the specified owned phi argument.
+///
+/// If the visitor returns false, stops visiting and returns false.  Otherwise,
+/// returns true.
+bool visitReborrowsOfOwnedPhiArgument(SILPhiArgument *arg,
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply include/swift/SIL/OwnershipUtils.h]: Great point!  Renamed and dropped the assertion.

File: lib/SIL/Utils/PrunedLiveness.cpp [hunk 4]
```diff
@@ -106,7 +106,8 @@ void PrunedLiveness::updateForUse(SILInstruction *user, bool lifetimeEnding) {
 }
 
 bool PrunedLiveness::updateForBorrowingOperand(Operand *op) {
-  assert(op->getOperandOwnership() == OperandOwnership::Borrow);
+  assert(op->getOperandOwnership() == OperandOwnership::Borrow ||
+         op->getOperandOwnership() == OperandOwnership::Reborrow);
```
[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SIL/Utils/PrunedLiveness.cpp]: Removed this bit, now that we're adding the adjacent reborrow phis to the def use worklist.

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Review by atrick — APPROVED** [review]
Awesome!

**Comment by nate-chandler** [comment]
@swift-ci please test macOS platform

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test

**Comment by nate-chandler** [comment]
@swift-ci please test macOS platform

**Comment by nate-chandler** [comment]
@swift-ci please test linux platform

**Pull request merged**